aspnet/mvc/views/working-with-forms.rst

```diff
@@ -0,0 +1,621 @@
+|[Url]                          |  type="url"        |
++-------------------------------+--------------------+  
+|[HiddenInput]                  |  type="hidden"     |
++-------------------------------+--------------------+  
+|[Phone]                        |  type="tel"        |
++-------------------------------+--------------------+   
+|[DataType(DataType.Password)]  |  type="password"   |
++-------------------------------+--------------------+  
+|[DataType(DataType.Date)]      |  type="date"       |
++-------------------------------+--------------------+  
+|[DataType(DataType.Time)]      |  type="time"       |
++-------------------------------+--------------------+  
+ 
+Sample: 
+ 
+.. literalinclude::  forms/sample/final/ViewModels/RegisterViewModel.cs
+  :language: c#
+
+.. literalinclude::  forms/sample/final/Views/Demo/RegisterInput.cshtml
+  :language: HTML
+
+The code above generates the following HTML:
+
+.. code-block:: HTML
+
+    <form method="post" action="/Demo/RegisterInput">
+      Email:  
+      <input type="email" data-val="true" 
+             data-val-email="The Email Address field is not a valid e-mail address." 
+             data-val-required="The Email Address field is required." 
+             id="Email" name="Email" value="" /> <br>
+      Password: 
+      <input type="password" data-val="true" 
+             data-val-required="The Password field is required." 
+             id="Password" name="Password" /><br>
+      <button type="submit">Register</button>
+    <input name="__RequestVerificationToken" type="hidden" value="<removed for brevity>" />
+  </form>
+
+The data annotations applied to the ``Email`` and ``Password`` properties generate metadata on the model. The Input Tag Helper consumes the model metadata and produces `HTML5 <https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5>`__ ``data-val-*`` attributes (see :doc:`/mvc/models/validation`). These attributes describe the validators to attach to the input fields. This provides unobtrusive HTML5 and `jQuery <https://jquery.com/>`__ validation. The unobtrusive attributes have the format ``data-val-rule="Error Message"``, where rule is the name of the validation rule (such as ``data-val-required``, ``data-val-email``, ``data-val-maxlength``, etc.) There are also attributes of the form ``data-val-ruleName-argumentName="argumentValue"``, for example, ``data-val-maxlength-max="1024"``.  If an error message is provided in the attribute, it is displayed as the value for the ``data-val-rule`` attribute. 
+
+Expression names
+^^^^^^^^^^^^^^^^^^^^^
+
+The ``asp-for`` attribute value is a `ModelExpression <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Rendering/ModelExpression/index.html>`__ and the right hand side of a lambda expression. Therefore, ``asp-for="Property1"`` becomes ``m => m.Property1`` in the generated code which is why you don't need to prefix with ``Model``.
```

Comment:
Per my comment at https://github.com/aspnet/Razor/issues/741#issuecomment-217269022, I would not show these examples with defining a new model within the page and then using it as `@model` in an expression. This usage is rarely seen in the wild so I want to avoid people copy/pasting from this sample.

Instead, just use the model that came from the controller action via `@Model` (uppercase `Model`).